### ISSUE TITLE
Slow performance of GIT 2.29.x, 2.28.x is just fine

### ISSUE THREAD

**Issue by basrieter** [git-for-windows/git#2879]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.28.0.windows.1
cpu: x86_64
built from commit: 77982caf269b7ee713a76da2bcf260c34d3bf7a7
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.18363.1139]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"

Editor Option: Notepad++
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Disabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

_None_

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git --version
```
But basically all commands

 - What did you expect to occur after running these commands?

Fast response.

 - What actually happened instead?

Git 2.29 (.0, .1 and .2) takes a long time to show any output. On git 2.28 a `git --version` would produce a nearly instant result, with 2.29 the result takes about 2 seconds. Any application that uses git (such as IntelliJ or PyCharm) take ages to perform git actions, as they fire a lot of git commands on the background.

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Independent of a repository.

**Comment by dscho** [comment]
Maybe https://github.com/git-for-windows/git/wiki/Diagnosing-performance-issues#trace-executions-in-the-bash-startup helps you diagnose this in more detail?

**Comment by basrieter** [comment]
So I checked some more: Seems like F-Secure does not like GIT 2.29.x and is slowing it down. I used Sysinternals ProcMon to check what was going on with `git --version` and I notice a waiting on file close and thread close/exit:

```
17:06:05.4721842	git.exe	15136	QueryNameInformationFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Name: \Program Files\Git\mingw64\bin\git.exe
17:06:05.4723488	git.exe	15136	CreateFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Desired Access: Generic Read, Disposition: Open, Options: Synchronous IO Non-Alert, Attributes: n/a, ShareMode: Read, Write, AllocationSize: n/a, OpenResult: Opened
17:06:05.4726728	git.exe	15136	ReadFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Offset: 0, Length: 1,048,576, Priority: Normal
17:06:05.4749711	git.exe	15136	ReadFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Offset: 1,048,576, Length: 1,048,576
17:06:05.4773334	git.exe	15136	ReadFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Offset: 2,097,152, Length: 1,048,576
17:06:05.4796945	git.exe	15136	ReadFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Offset: 3,145,728, Length: 174,640
17:06:05.4801926	git.exe	15136	ReadFile	C:\Program Files\Git\mingw64\bin\git.exe	END OF FILE	Offset: 3,320,368, Length: 1,048,576
17:06:05.4803178	git.exe	15136	CloseFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	
17:06:06.1913161	git.exe	15136	QueryNameInformationFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Name: \Program Files\Git\mingw64\bin\git.exe
17:06:06.1917841	git.exe	15136	QueryNameInformationFile	C:\Program Files\Git\mingw64\bin\git.exe	SUCCESS	Name: \Program Files\Git\mingw64\bin\git.exe
```
and 
```
17:06:06.2030663	git.exe	15136	CloseFile	C:\Users\bas\.gitconfig	SUCCESS	
17:06:06.2041436	git.exe	15136	Thread Exit		SUCCESS	Thread ID: 3656, User Time: 0.0000000, Kernel Time: 0.0000000
17:06:06.2042416	git.exe	15136	Thread Exit		SUCCESS	Thread ID: 1620, User Time: 0.0000000, Kernel Time: 0.0000000
17:06:06.2042471	git.exe	15136	Thread Exit		SUCCESS	Thread ID: 11324, User Time: 0.0000000, Kernel Time: 0.0000000
17:06:06.2042538	git.exe	15136	Thread Exit		SUCCESS	Thread ID: 20816, User Time: 0.0000000, Kernel Time: 0.0000000
17:06:06.8943903	git.exe	15136	RegOpenKey	HKLM\Software\Microsoft\Windows NT\CurrentVersion\GRE_Initialize	SUCCESS	Desired Access: Read
17:06:06.8944205	git.exe	15136	RegQueryValue	HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion\GRE_Initialize\DisableMetaFiles	NAME NOT FOUND	Length: 20
17:06:06.8944480	git.exe	15136	RegCloseKey	HKLM\SOFTWARE\Microsoft\Windows NT\CurrentVersion\GRE_Initialize	SUCCESS	
17:06:06.8945717	git.exe	15136	Thread Exit		SUCCESS	Thread ID: 13868, User Time: 0.0781250, Kernel Time: 1.3593750
17:06:06.8953740	git.exe	15136	Process Exit		SUCCESS	Exit Status: 0, User Time: 0.0781250 seconds, Kernel Time: 1.3593750 seconds, Private Bytes: 3,248,128, Peak Private Bytes: 3,919,872, Working Set: 5,984,256, Peak Working Set: 6,615,040
```

This causes a 1 second delay, that is not present in GIT 2.28. If an application fires multiple git commands, this might actually add up to a long delay: I notice 20-30 seconds on history listing in PyCharm.

**Comment by basrieter** [comment]
If I exclude "c:\Program Files\Git\mingw64\bin\git.exe" the slowdown is gone.

**Comment by dscho** [comment]
Thank you for digging in further!

Sadly, this is not the first time anti-malware causes trouble with Git for Windows, not by a far stretch, and I fear it won't be the last time, either :-(

**Comment by basrieter** [comment]
Would it be possible to find the root-cause for this? In other words: what changed in GIT for Windows that might cause this? If I rollback to 2.28 all is fine.

**Comment by dscho** [comment]
> Would it be possible to find the root-cause for this? In other words: what changed in GIT for Windows that might cause this? If I rollback to 2.28 all is fine.

of course it is, but only for someone experiencing the slow-down (I, for one, do not).

In a similar (now closed) ticket, the root cause was tracked back to F-Secure being a bit too eager with v2.29.0.

You could also bisect through the [snapshots](https://wingit.blob.core.windows.net/files/index.html) to pin-point the commit with which the slow-down happens.

**Comment by PhilipOakley** [comment]
On 01/11/2020 06:59, Johannes Schindelin wrote:
>
>     Would it be possible to find the root-cause for this? In other
>     words: what changed in GIT for Windows that might cause this? If I
>     rollback to 2.28 all is fine.
>
> of course it is, but only for someone experiencing the slow-down (I,
> for one, do not).
>
> In a similar (now closed) ticket, the root cause was tracked back to
> F-Secure being a bit too eager with v2.29.0.
>
> You could also bisect through the snapshots
> <https://wingit.blob.core.windows.net/files/index.html> to pin-point
> the commit with which the slow-down happens.
>
@basrieter,

You may find that it magically solves itself once the virus scanner
folks (if that's the cause for you) determine that there is a new
version of Git with a slightly updated signature, and that they push out
an updated signature file to their customers, and then the commercial
customers push it to their users.

**Comment by lanfeust69** [comment]
I have the same kind of issue with 2.29, and I have a very strong suspicion it is related to the Symantec cr*p that runs on our PC, and that can't be disabled, even with admin rights, and even temporarily to confirm that it is the cause (sigh).

In my case, this is most notable during a fetch over ssh on our (quite big) main repo, where 2.29 may take up to 10 minutes (of kernel CPU time) for fetching nothing, while 2.28 takes a few seconds.

Even though I saw the problem almost immediately after the release, I haven't found the time to dig more. I just kept a portable git 2.28 around for these fetches...

**Comment by helixhorned** [comment]
I had the same issues since the version update mentioned in the issue title, but so far I could tolerate it because the delays were about a few seconds at most.

Today, I observed something which makes me suggest a different theory to the ones proposed here. Briefly, could it not be that the delays are due to attempts to access the network?

It's just that, under a certain network configuration A, for me, Git Bash today basically crawled to a halt, but I observed that under this configuration, anything involving network access did. When I switched to network configuration B, both Git bash and the normal network usage were fine again.

**Comment by dscho** [comment]
> could it not be that the delays are due to attempts to access the network?

Absolutely. This might be a network call (e.g. to your domain controller). I lack the time to investigate this angle, but maybe you can?

**Comment by lanfeust69** [comment]
At least in my case, I'm pretty sure it isn't something network-related.

The slowdown occurs *after* the network part is done, where ProcExp shows one core churning at 100% in kernel mode, with stack traces within Symantec dll.

Moreover, a succession of fetches (so basically no-ops at that time) are perfectly consistent, with 2.29 (or 2.30) taking ~10 minutes, whereas 2.28 takes less than 30 seconds.

Regarding the possibility to bisect, unfortunately I think we are in the realm @PhilipOakley mentioned : my local build of 2.28 displays the same slow-down as official 2.29 and 2.30...

**Comment by helixhorned** [comment]
> > could it not be that the delays are due to attempts to access the network?
> 
> Absolutely. This might be a network call (e.g. to your domain controller). I lack the time to investigate this angle, but maybe you can?

I'm afraid that's not realistic. I'm using Git for Windows (including its version of Bash) at work and then only when I'm in Windows proper. Also, most of the time, the delay is noticeable but within agreeable bounds.

However, re-reading the original post now makes me wonder if what I'm seeing is the same issue. @basrieter mentions using Git commands under `cmd.exe`, while what I'm experiencing is even delays between **presses of the Enter key** (with no command) in the Bash provided by this project, and this is what my observation referrred to -- even those ones got worse under a bad network state.

So, I'm sorry if I posted in the wrong issue without careful reading. I got here by web search and it sounded relevant at first :-). It is noteable though that, if memory serves me right, the behavior I'm seeing started occurring around the same version as mentioned by the OP.
